caiman/source_extraction/cnmf/lazy_arrays.py

```diff
@@ -0,0 +1,444 @@
+        prods = list()
+        for t, s in iter_product([temporal_min, temporal_max], [spatial_min, spatial_max]):
+            _p = np.multiply(t, s)
+            prods.append(np.nanmin(_p))
+            prods.append(np.nanmax(_p))
+
+        self._max = np.max(prods)
+        self._min = np.min(prods)
+
+        # get the mean of each temporal component
+        temporal_mean = np.nanmean(self.temporal, axis=1)
+
+        # get the standard deviation of each temporal component
+        temporal_std = np.nanstd(self.temporal, axis=1)
+
+        # compute mean, max, min, and std projection images
+        self._mean_image = (self.spatial @ temporal_mean).reshape(frame_dims, order="F")
+        self._max_image = (self.spatial @ temporal_max).reshape(frame_dims, order="F")
+        self._min_image = (self.spatial @ temporal_min).reshape(frame_dims, order="F")
+        self._std_image = (self.spatial @ temporal_std).reshape(frame_dims, order="F")
+
+    @property
+    def spatial(self) -> np.ndarray:
+        return self._spatial
+
+    @property
+    def temporal(self) -> np.ndarray:
+        return self._temporal
+
+    @property
+    def n_components(self) -> int:
+        return self._spatial.shape[1]
+
+    @property
+    def n_frames(self) -> int:
+        return self._temporal.shape[1]
+
+    @property
+    def shape(self) -> Tuple[int, int, int]:
+        return self._shape
+
+    @property
+    def min(self) -> float:
+        return self._min
+
+    @property
+    def max(self) -> float:
+        return self._max
+
+    @property
+    def mean_image(self) -> np.ndarray:
+        """mean projection image"""
+        return self._mean_image
+
+    @property
+    def max_image(self) -> np.ndarray:
+        """max projection image"""
+        return self._max_image
+
+    @property
+    def min_image(self) -> np.ndarray:
+        """min projection image"""
+        return self._min_image
+
+    @property
+    def std_image(self) -> np.ndarray:
+        """standard deviation projection image"""
+        return self._std_image
+
+    def _compute_at_indices(self, indices: Union[int, Tuple[int, int]]) -> np.ndarray:
+        rcm = (self.spatial @ self.temporal[:, indices]).reshape(
+            self.shape[1:] + (-1,), order="F"
+        ).transpose([2, 0, 1])
+
+        if rcm.shape[0] == 1:
+            return rcm[0]  # 2d single frame
+        else:
+            return rcm
+
+    def __repr__(self):
+        r = super().__repr__()
+        return f"{r}" \
+               f"n_components: {self.n_components}"
+
+    def __eq__(self, other):
+        if not isinstance(other, self.__class__):
+            raise TypeError(f"cannot compute equality against types that are not {self.__class__.__name__}")
+
+        if (self.spatial == other.spatial) and (self.temporal == other.temporal):
+            return True
+        else:
+            return False
+
+
+# implementation for reconstructed background is identical
```

Comment:
background is messy b/c of CNMFE this is the major remaining hurdle I think.  But of course Eftychios and Co already foresaw this and implemented it with `compute_background()` :smile: 

So I think in general this would be very useful to have as its own thing.

Reply:
Hmmm- actually thinking about it a bit more.  Ignoring ssub_B for now assuming subsampling is 1 for the background fitting, the background in the CNMFE model is:

    background_model = b0 + W(Yr - AC - b0)

This is basically combining equations 5 and 8 from the CNMFE paper. 

Hmm maybe thought: all the things in the parens are do-able: b0 is easy (because it is multiplied by a vector of 1's of length T). AC is easy you already do that. Yr is already memmapped.  So basically subtraction operation inside parens, with some mult, and then matrix mult of W (which is huge) by that thing in the parens (W is like dims x dims). 

I think it can be done. And then we can get around the caveat in the code for compute background it throws a warning:

     logging.warning("Computing the full background has big memory requirements!")

:smile: 

It would take some more figuring out details with all this, and especially the `ssub_B` bits which I left out, but I think this should be doable with the framework you've built!

Reply:
> Yr is already memmapped

Always?? Is this a CNMFE specific thing?

> Once that operation is done, then matrix mult of W (which is a huge matrix). The nice thing is we wouldn't need to do all columns of the things inside the parens because lazy array.

Yes, `W` can have shape that is large, but still in the order of megabytes so it's not that bad. W is dense, the others are still sparse (right?) so it's should still be fast. 

> I think there are repercussions of this lazy array stuff that we just don't appreciate, tbh, that will be really cool as we work out the implications of the math in the models etc

Right now some of the most useful things are the projections, but there definitely must be other uses of lazy computing we can take advantage of. I wonder how much of dask's optimization engine is smart enough to catch all these things. For example, if it is able to figure out the standard deviation of the residuals without calculating the full thing that will be amazing.

Reply:
I am not aware of `Yr` being used in a non-memmapped way in Caiman, but I may be wrong. We are pretty careful with that because it is huge.